File: library/src/main/java/org/xmtp/android/library/frames/FramesClient.kt
package org.xmtp.android.library.frames

import android.util.Base64
import com.google.protobuf.kotlin.toByteString
import org.xmtp.android.library.Client
import org.xmtp.android.library.XMTPException
import org.xmtp.android.library.frames.FramesConstants.PROTOCOL_VERSION
import org.xmtp.android.library.hexToByteArray
import org.xmtp.android.library.toHex
import org.xmtp.proto.message.contents.Frames.FrameAction
import org.xmtp.proto.message.contents.Frames.FrameActionBody
import java.security.MessageDigest
import java.util.Date

class FramesClient(private val xmtpClient: Client, var proxy: OpenFramesProxy = OpenFramesProxy()) {

    suspend fun signFrameAction(inputs: FrameActionInputs): FramePostPayload {
        val opaqueConversationIdentifier = buildOpaqueIdentifier(inputs)
        val frameUrl = inputs.frameUrl
        val buttonIndex = inputs.buttonIndex
        val inputText = inputs.inputText
        val state = inputs.state
        val now = Date().time / 1_000
        val frameActionBuilder = FrameActionBody.newBuilder().also { frame ->
            frame.frameUrl = frameUrl
            frame.buttonIndex = buttonIndex
            frame.opaqueConversationIdentifier = opaqueConversationIdentifier
            frame.unixTimestamp = now.toInt()
            if (inputText != null) {
                frame.inputText = inputText
            }
            if (state != null) {
                frame.state = state
            }
        }

        val toSign = frameActionBuilder.build()
        val signedAction = Base64.encodeToString(buildSignedFrameAction(toSign), Base64.NO_WRAP)

        val untrustedData = FramePostUntrustedData(
            frameUrl,
            now,
            buttonIndex,
            inputText,
            state,
            xmtpClient.address,
            opaqueConversationIdentifier,
            now.toInt()
        )
        val trustedData = FramePostTrustedData(signedAction)

        return FramePostPayload("xmtp@$PROTOCOL_VERSION", untrustedData, trustedData)
    }

    private fun signDigest(message: String): ByteArray {
        return xmtpClient.signWithInstallationKey(message)
    }

    private fun buildSignedFrameAction(actionBodyInputs: FrameActionBody): ByteArray {
        val digest = sha256(actionBodyInputs.toByteArray()).toHex()
        val signature = signDigest(digest)

        val frameAction = FrameAction.newBuilder().also {
            it.actionBody = actionBodyInputs.toByteString()
            it.installationSignature = signature.toByteString()
            it.installationId = xmtpClient.installationId.hexToByteArray().toByteString()
            it.inboxId = xmtpClient.inboxId
        }.build()

        return frameAction.toByteArray()
    }

    private fun buildOpaqueIdentifier(inputs: FrameActionInputs): String {
        return when (inputs.conversationInputs) {
            is ConversationActionInputs.Group -> {
                val groupInputs = inputs.conversationInputs.inputs
                val combined = groupInputs.groupId + groupInputs.groupSecret
                val digest = sha256(combined)
                Base64.encodeToString(digest, Base64.NO_WRAP)
            }

            is ConversationActionInputs.Dm -> {
                val dmInputs = inputs.conversationInputs.inputs
                val conversationTopic =
                    dmInputs.conversationTopic ?: throw XMTPException("No conversation topic")
                val combined =
                    conversationTopic.lowercase() + dmInputs.participantAccountAddresses.map { it.lowercase() }
                        .sorted().joinToString("")
                val digest = sha256(combined.toByteArray())
                Base64.encodeToString(digest, Base64.NO_WRAP)
            }
        }
    }

    private fun sha256(input: ByteArray): ByteArray {
        val digest = MessageDigest.getInstance("SHA-256")
        return digest.digest(input)
    }
}

File: Sources/XMTPiOS/Frames/FramesClient.swift
//
//  FramesClient.swift
//  
//
//  Created by Alex Risch on 3/28/24.
//

import Foundation
import LibXMTP

public typealias FrameActionBody = Xmtp_MessageContents_FrameActionBody
public typealias FrameAction = Xmtp_MessageContents_FrameAction

enum FramesClientError: Error {
    case missingConversationTopic
    case missingTarget
    case readMetadataFailed(message: String, code: Int)
    case postFrameFailed(message: String, code: Int)
}

public class FramesClient {
    var xmtpClient: Client
    public var proxy: OpenFramesProxy

    public init(xmtpClient: Client, proxy: OpenFramesProxy? = nil) {
        self.xmtpClient = xmtpClient
        self.proxy = proxy ?? OpenFramesProxy()
    }

    public func signFrameAction(inputs: FrameActionInputs) async throws -> FramePostPayload {
        let opaqueConversationIdentifier = try self.buildOpaqueIdentifier(inputs: inputs)
        let frameUrl = inputs.frameUrl
        let buttonIndex = inputs.buttonIndex
        let inputText = inputs.inputText ?? ""
        let state = inputs.state ?? ""
        let now = Date().timeIntervalSince1970
        let timestamp = now

        var toSign = FrameActionBody()
        toSign.frameURL = frameUrl
        toSign.buttonIndex = buttonIndex
        toSign.opaqueConversationIdentifier =  opaqueConversationIdentifier
        toSign.timestamp = UInt64(timestamp)
        toSign.inputText = inputText
        toSign.unixTimestamp = UInt32(now)
        toSign.state = state

        let signedAction = try await self.buildSignedFrameAction(actionBodyInputs: toSign)

        let untrustedData = FramePostUntrustedData(
            url: frameUrl, timestamp: UInt64(now), buttonIndex: buttonIndex, inputText: inputText, state: state, walletAddress: self.xmtpClient.address, opaqueConversationIdentifier: opaqueConversationIdentifier, unixTimestamp: UInt32(now)
        )


        let trustedData = FramePostTrustedData(messageBytes: signedAction.base64EncodedString())

        let payload = FramePostPayload(
            clientProtocol: "xmtp@\(PROTOCOL_VERSION)", untrustedData: untrustedData, trustedData: trustedData
        )

        return payload
    }
    
    private func signDigest(digest: Data) async throws -> Signature {
        let key = self.xmtpClient.keys.identityKey
        let privateKey = try PrivateKey(key)
        let signature = try await privateKey.sign(Data(digest))
        return signature
    }
    
    private func getPublicKeyBundle() async throws -> PublicKeyBundle {
        let bundleBytes = self.xmtpClient.publicKeyBundle;
        return try PublicKeyBundle(bundleBytes);
      }
    
    private func buildSignedFrameAction(actionBodyInputs: FrameActionBody) async throws -> Data {

        let digest = sha256(input: try actionBodyInputs.serializedData())
        let signature = try await self.signDigest(digest: digest)

        let publicKeyBundle = try await self.getPublicKeyBundle()
        var frameAction = FrameAction()
        frameAction.actionBody = try actionBodyInputs.serializedData()
        frameAction.signature = signature
        frameAction.signedPublicKeyBundle = try SignedPublicKeyBundle(publicKeyBundle)
        
        return try frameAction.serializedData()
    }
    
    private func buildOpaqueIdentifier(inputs: FrameActionInputs) throws -> String {
        switch inputs.conversationInputs {
        case .group(let groupInputs):
            let combined = groupInputs.groupId + groupInputs.groupSecret
            let digest = sha256(input: combined)
            return digest.base64EncodedString()
        case .dm(let dmInputs):
            guard let conversationTopic = dmInputs.conversationTopic else {
                throw FramesClientError.missingConversationTopic
            }
            guard let combined = (conversationTopic.lowercased() + dmInputs.participantAccountAddresses.map { $0.lowercased() }.sorted().joined()).data(using: .utf8) else {
                throw FramesClientError.missingConversationTopic
            }
            let digest = sha256(input: combined)
            return digest.base64EncodedString()
        }
    }

}
